src/dialer/index.js

```diff
@@ -0,0 +1,129 @@
+'use strict'
+
+const multiaddr = require('multiaddr')
+const errCode = require('err-code')
+const AbortController = require('abort-controller')
+const anySignal = require('any-signal')
+const debug = require('debug')
+const log = debug('libp2p:dialer')
+log.error = debug('libp2p:dialer:error')
+const { DialRequest } = require('./dial-request')
+
+const { codes } = require('../errors')
+const {
+  DIAL_TIMEOUT,
+  MAX_PARALLEL_DIALS,
+  MAX_PER_PEER_DIALS
+} = require('../constants')
+
+class Dialer {
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {TransportManager} options.transportManager
+   * @param {Peerstore} peerStore
+   * @param {number} options.concurrency Number of max concurrent dials. Defaults to `MAX_PARALLEL_DIALS`
+   * @param {number} options.timeout How long a dial attempt is allowed to take. Defaults to `DIAL_TIMEOUT`
+   */
+  constructor ({
+    transportManager,
+    peerStore,
+    concurrency = MAX_PARALLEL_DIALS,
+    timeout = DIAL_TIMEOUT,
+    perPeerLimit = MAX_PER_PEER_DIALS
+  }) {
+    this.transportManager = transportManager
+    this.peerStore = peerStore
+    this.concurrency = concurrency
+    this.timeout = timeout
+    this.perPeerLimit = perPeerLimit
+    this.tokens = [...new Array(concurrency)].map((_, index) => index)
+    this.pendingDials = new Set()
+  }
+
+  /**
+   * Connects to the first success of a given list of `Multiaddr`. `addrs` should
+   * include the id of the peer being dialed, it will be used for encryption verification.
+   *
+   * @param {Array<Multiaddr>|Multiaddr} addrs
+   * @param {object} [options]
+   * @param {AbortSignal} [options.signal] An AbortController signal
+   * @returns {Promise<Connection>}
+   */
+  async connectToMultiaddr (addrs, options = {}) {
+    if (!Array.isArray(addrs)) addrs = [multiaddr(addrs)]
+
+    const dialAction = (addr, options) => {
+      if (options.signal.aborted) throw errCode(new Error('already aborted'), 'ERR_ALREADY_ABORTED')
```

Comment:
```suggestion
      if (options && options.signal && options.signal.aborted) throw errCode(new Error('already aborted'), 'ERR_ALREADY_ABORTED')
```

Reply:
This isn't necessary. `DialRequest` will **always** call `dialAction` with a signal.

Reply:
...then it is not an option 😜

Reply:
Well, technically an option isn't necessarily optional, and we are internally requiring the optional values of `Transport.dial()`. English is fun.